### PR TITLE
add summary file

### PR DESCRIPTION
**Pull request by danielecook**
Hello - not sure if you are open to PRs but I've thrown together one that adds a summary file for the whole genome. I'm still working out how to do this at the region level.

The output for whole-genome would be: `$prefix.mosdepth.global.summary.txt`

An example output looks like this:

| contig   |     bases |   mean |   median |     sd |   min |   max |
|:---------|----------:|-------:|---------:|-------:|------:|------:|
| I        |  15072435 |  20.26 |       18 |  45.87 |     0 |  2746 |
| II       |  15279422 |  19.83 |       19 |  18.83 |     0 |  1554 |
| III      |  13783802 |  18.96 |       18 |  15.09 |     0 |  1626 |
| IV       |  17493830 |  19.49 |       18 |  18.29 |     0 |  1871 |
| V        |  20924181 |  20.04 |       19 |  20.53 |     0 |  2866 |
| X        |  17718943 |  23.81 |       22 |  25.39 |     0 |  3937 |
| MtDNA    |     13795 | 751.74 |      744 | 298.94 |     0 |  1611 |
| total    | 100286408 |  20.56 |       19 |  27.39 |     0 |  3937 |

The mean/median include depths of 0. Still have to validate/check but perhaps this is useful?

### CODE REVIEW

**Comment by brentp** [comment]
very cool! I'll accept this with tests and support for regions files.
can you also verify that it does not affect runtime more than 1 or 2%?

**Comment by danielecook** [comment]
Sure I’ll see what I can do.

**Comment by danielecook** [comment]
I've added regions and had a look at the benchmarks. Adding a summary like this definitely slows things down by more than 1-2%. I am using the `RunningStat` object for mean/sd/min/max and a `CountTable` for the median. You may know of a more efficient method.

I often want to know mean and median depth of sequencing runs and this implementation runs 10x faster than GATK and would still presumably run faster than samtools depth. 

Perhaps it could only run with an optional parameter (`--summary`)? I'll leave it up to you whether you want to incorporate this. I will probably use it regardless. Still happy to write the tests if you are open to merging.

Here is what I am seeing with WGS from _C. elegans_ and a bedfile of genes and 4 threads:

__With Summary__
```
real	1m23.959s
user	1m39.243s
sys	0m1.888s
```

__Original__
```
real	1m4.765s
user	1m19.940s
sys	0m1.878s
```

And example results with region info:

| contig       |     bases |   mean |   median |     sd |   min |   max |
|:-------------|----------:|-------:|---------:|-------:|------:|------:|
| I            |  15072435 |  24.68 |       22 |  55.32 |     0 |  3363 |
| I_region     |  10726160 |  24.65 |       22 |  54.58 |     0 |  3363 |
| II           |  15279422 |  24.15 |       23 |  24.69 |     0 |  2088 |
| II_region    |   9812203 |  24.58 |       24 |  15.73 |     0 |   928 |
| III          |  13783802 |  23.12 |       22 |  19.75 |     0 |  1841 |
| III_region   |   9636582 |  23.28 |       22 |  14.61 |     0 |  1629 |
| IV           |  17493830 |  23.79 |       22 |  24.27 |     0 |  2941 |
| IV_region    |  10455087 |  24.34 |       23 |  12.7  |     0 |   567 |
| V            |  20924181 |  24.38 |       23 |  26.17 |     0 |  3728 |
| V_region     |  12405213 |  25.04 |       24 |  15.3  |     0 |  1086 |
| X            |  17718943 |  28.99 |       27 |  33.64 |     0 |  5786 |
| X_region     |   9872441 |  30.12 |       29 |  13.39 |     0 |   840 |
| MtDNA        |     13795 | 953.07 |      931 | 379.29 |     0 |  2164 |
| MtDNA_region |     12085 | 993.5  |      954 | 354.45 |   275 |  2164 |
| total        | 100286408 |  25.05 |       23 |  34.56 |     0 |  5786 |
| total_region |  62919771 |  25.5  |       24 |  29.81 |   275 |  3363 |

On the plus side, I identified a bug (https://github.com/nim-lang/Nim/commit/1332f649b2e32a84fb643b580750e2f554eb8a4f) in the `runningStat` object (adding two together should produce the minimum of both, but was producing the max, hence the error in the total min for `total_region`). This should be fixed when nim gets updated.

**Comment by brentp** [comment]
hmm. that's more than I'd like and I'd prefer not to add another command-line option. This is useful information to output, but I think it can be done with little overhead.

can you see where the additional time is being spent? e.g. if you temporarily change the median implementation to just `return 4` or something, does the difference drop?  The median coverage is already calculated for both regions and globals, so we could probably avoid that. It could also be related to the use of globals, which, don't do well for performance in nim IIUC.

I think I would write a function that takes a depth array and an optional list of regions and returns 2 RunningStat objects, 1 for the full array, and 1 for the regions (if they were provided).

Also, for regions, it should report as a BED file with the original 1st 3 columns (I assume you're currently getting the name like `I_region` from the 4th column?).

**Comment by danielecook** [comment]
I can tell you calculating median accounts for about 1/3 of the added time.

I’ll play around with the globals and see if there are other ways to speed this up including retrieving the median from elsewhere. Thanks for your feedback.

**Comment by brentp** [comment]
ok. great. thanks for persisting. I do think that having the function I described will make it easy to test. If you implement without the median and the performance is relatively unaffected, then I will merge and I can add the median from the current calculation after that.

**Comment by danielecook** [comment]
Regarding the question of regions - `I_region` is not from the 4th column, it is reporting the average/sd/min/max/median for all regions on chromosome `I`. It would be possible to output the runningStat for each individual region as well.

I'm primarily interested in the mean/median depth when looking at whole exome data and really only need the `total_region` row, so aggregating by chromosome may not be necessary in the summary output. (The bam I am using here is WGS, not WES but is useful for testing)

Would it be more preferable to have the following (where RunningStat is count/mean/sd/min/max + median)?:
    - RunningStat data for each individual region
    - RunningStat of all regions
    - RunningStat of whole genome

**Comment by danielecook** [comment]
ok cool I'll take a crack at implementing it with the function you suggested. Thanks!

**Comment by brentp** [comment]
oh I see, regarding regions. that's better as you have it. we certainly don't need stats for every region in the bed file.

**Comment by danielecook** [comment]
After experimenting for a while I realize that the usage of `RunningStat` is contributing to the slowdown. Calculating the cumulative mean, min and max is trivial. However, calculating the standard dev. (+variance and other items) adds considerable overhead. This is because `RunningStat` must recompute the mean for every value that is pushed. See [Welford's online algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance) and [nim RunningStat](https://github.com/nim-lang/Nim/blob/master/lib/pure/stats.nim#L80) for more details. The only way to calc. std. dev. is to store the entire genome or use a runningStat. 

Fortunately I don't need the std. dev. So I used a much simpler implementation that is much more efficient, but still falls short a little short of 1-2% slowdown. 

Using this command on a slightly faster computer than before:
```
time ./mosdepth --threads 4 --fast-mode --by genes.bed test JU3224.bam
```

__New implementation__
```
real	1m2.678s
user	1m18.566s
sys	0m1.832s
```

__Original implementation__
```
real	0m58.042s
user	1m14.378s
sys	0m1.881s
```

Same output, except median is missing and would be useful:

| chrom        |    length |      bases |   mean |   min |   max |
|:-------------|----------:|-----------:|-------:|------:|------:|
| I            |  15072435 |  371921528 |  24.68 |     0 |  3363 |
| I_region     |  10726160 |  264370372 |  24.65 |     0 |  3363 |
| ... |
| MtDNA_region |     12085 |   12006421 | 993.5  |   275 |  2164 |
| total        | 100286408 | 2512543615 |  25.05 |     0 |  5786 |
| total_region |  62919771 | 1604305947 |  25.5  |     0 |  3363 |

Let me know if you are still interested in incorporating this. I am happy to make further changes.

Thanks!

File: mosdepth.nim [hunk 1]
```diff
@@ -125,7 +151,7 @@ iterator gen_quantized(quants: seq[int], arr: coverage_t): depth_s {.inline.} =
         yield (last_pos, pos, lookup[last_quantized])
       last_quantized = quantized
       last_pos = pos
-    if last_quantized != -1 and last_pos < arr.high and last_quantized < len(lookup):
+    if last_quantized != -1 and last_pos < arr.high:
```
Comment (brentp) [inline mosdepth.nim]: make sure you dont' change this. It was a recent bug-fix.

File: mosdepth.nim [hunk 2]
```diff
@@ -25,9 +28,32 @@ type
...
+  depth_stat = object
+    cum_length: uint32
+    cum_depth: uint64
+    min_depth: uint32
+    max_depth: uint32
 
   coverage_t = seq[int32]
 
+proc newDepthStat(d: coverage_t): depth_stat =
+  return depth_stat(cum_length: len(d).uint32,
```
Comment (brentp) [inline mosdepth.nim]: use int for `cum_length` or we won't be able to handle large choms/genomes.

File: mosdepth.nim [hunk 3]
```diff
@@ -667,7 +732,7 @@ when(isMainModule):
   when not defined(release) and not defined(lto):
     stderr.write_line "[mosdepth] WARNING: built in debug mode; will be slow"
 
-  let version = "mosdepth 0.2.6"
+  let version = "mosdepth 0.2.5"
```
Comment (brentp) [inline mosdepth.nim]: make sure you rebase on top of current master so you don't do stuff like lower the version.

File: mosdepth.nim (hunk 2 above)
Comment (danielecook) [reply mosdepth.nim]: Got it! Thanks!

File: mosdepth.nim (hunk 1 above)
Comment (danielecook) [reply mosdepth.nim]: Thanks I put those fixes back. Did not realize you had updated things.

File: mosdepth.nim (hunk 3 above)
Comment (danielecook) [reply mosdepth.nim]: Fixed

File: mosdepth.nim (hunk 2 above)
Comment (danielecook) [reply mosdepth.nim]: Fixed thanks

**Comment by danielecook** [comment]
> how many does your `genes.bed` have?

`genes.bed` has 21,655 intervals.

In terms of tests, I am going to verify that results are identical to those found with other tools. I can add a test that verifies a mean from one of the test bams.

**Comment by brentp** [comment]
ok. can you add a test to tests/functional-tests.sh

a reasonable test would be to use an interval the full size of the chromosome and make sure it has a value equal to the chromosome.

File: mosdepth.nimble [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Package
 
-version       = "0.2.6"
+version       = "0.2.5"
```
Comment (brentp) [inline mosdepth.nimble]: don't forget version bump. you should rebase on master.

**Comment by brentp** [comment]
sorry, I let this fall off my radar. is this ready on your end?

**Comment by danielecook** [comment]
The ball is still in my court. Sorry - fell off my radar as well. I will try to take a look later today. The main thing is to just write a couple of tests.

**Comment by brentp** [comment]
ok. no worries. I just didn't want to be the bottleneck after all your work on this.

**Comment by danielecook** [comment]
I've added some functional tests and everything passes. That's a really cool little bash library for testing!

- It runs a little slower for me (`25.97` original v. `27.16` updated)
- I wasn't sure why, but the `coverage_t`/`arr` sequence is apparently 1 longer than the length of the chromosome. Does that sound right? I had to do this to get the correct chromosome length when outputting for the chromosome (but not for regions):

```
chrom_stat = newDepthStat(arr[0..<len(arr)-1])
```


Running these two commands gives identical averages: 

```
MOSDEPTH_PRECISION=8; ./mosdepth -x t tests/ovl.bam
samtools depth -a tests/ovl.bam  |  awk '{sum+=$3} END { print "Average = ",sum/NR}'
```

When comparing with samtools on real data I have for _C. elegans_ the averages were within `0.0136` of one another (using samtools with `-a` and mosdepth with `-x`). It seems there might be some differences small differences in terms of how the depth is calculated or perhaps something to do with secondary/supp reads possibly? Do you have any ideas?

You can reproduce with the code below. It prob. would take less than 15 mins to run.

```
wget https://s3.us-east-2.amazonaws.com/elegansvariation.org/bam/JU3224.bam
samtools index JU3224.bam

# mosdepth gives:
# Average =25.0537  sum = 2512543615 bases = 100286401 
MOSDEPTH_PRECISION=4; ./mosdepth -x t JU3224.bam

# samtools gives:
# Average =  25.0401  sum =  2511177329  bases =  100286401
samtools depth -a JU3224.bam  | awk '{sum+=$3} END { print "Average = ",sum/NR, " sum = ", sum, " bases = ", NR}'
```

So the bases counted are slightly diff: `2512543615` v. `2511177329`

Let me know if you find anything else. Thanks!

**Comment by brentp** [comment]
thanks very much!

**Comment by brentp** [comment]
there are many issues on the differences between samtools depth and mpileup and mosdepth. I am confident that mosdepth is doing the right thing and that I understand why they are different.

**Pull request merged**